sys/uartpb/UartProtobuf.cpp

```diff
@@ -53,49 +54,83 @@ void UartProtobuf::uart_rx_cb(uint8_t data)
 {
+    uuid_t uuid;
     msg_t msg;
     msg_t msg_queue[8];
     msg_init_queue(msg_queue, 8);
 
     while (1) {
+        // Wait for a new message
         msg_receive(&msg);
+
+        // Read and check message length
         uint32_t message_length = (uint32_t)msg.content.value;
         assert(message_length <= UARTPB_INPUT_MESSAGE_LENGTH_MAX);
-        read_buffer_.clear();
-        ringbuffer_get(&rx_buf_, (char *)read_buffer_.get_base64_data(), message_length);
-        size_t res = read_buffer_.base64_decode();
-        if (res > 0 && message_handler_) {
-            message_handler_(read_buffer_);
+
+        // Read uuid
+        ringbuffer_get(&rx_buf_, (char *)&uuid, sizeof(uuid_t));
+        message_length -= sizeof(uuid_t);
+
+        // Check a handler corresponding to the uuid is registered
+        if (message_handlers_.count(uuid) != 1) {
+            std::cout << "Unknown message uuid: " << uuid << std::endl;
+            continue;
         }
+
+        // Read Protobuf message if any
+        if (message_length > 0) {
+            read_buffer_.clear();
+            ringbuffer_get(&rx_buf_, (char *)read_buffer_.get_base64_data(), message_length);
+            size_t res = read_buffer_.base64_decode();
+            if (res <= 0) {
+                std::cout << "Failed to base64 decode Protobuf message (res = " << res <<  ")" << std::endl;
+                continue;
+            }
+        }
+
+        message_handlers_[uuid](message_length > 0 ? &read_buffer_ : nullptr);
     }
 }
-
-bool UartProtobuf::send_message(const EmbeddedProto::MessageInterface &message)
+bool UartProtobuf::send_message(uuid_t uuid, const EmbeddedProto::MessageInterface *message)
 {
     bool success = true;
+    size_t base64_size = 0;
+    char separator = '\n';
     mutex_.lock();
-    write_buffer_.clear();
-    auto serialization_status = message.serialize(write_buffer_);
-    if(EmbeddedProto::Error::NO_ERRORS != serialization_status) {
-        puts("Failed to serialize Protobuf message.");
-        success = false;
-    }
-    else {
-        size_t base64_size = write_buffer_.base64_encode();
-        if (base64_size == 0) {
-            puts("Failed to base64 encode Protobuf serialized message.");
+
+    if (message) {
+        write_buffer_.clear();
+        auto serialization_status = message->serialize(write_buffer_);
+        if(EmbeddedProto::Error::NO_ERRORS != serialization_status) {
+            puts("Failed to serialize Protobuf message.");
             success = false;
         }
         else {
-            uart_write(uart_dev_, write_buffer_.get_base64_data(), base64_size+1);
+            base64_size = write_buffer_.base64_encode();
+            if (base64_size == 0) {
+                puts("Failed to base64 encode Protobuf serialized message.");
+                success = false;
+            }
+        }
+    }
+    if (success) {
+        uart_write(uart_dev_, (uint8_t *)&uuid, sizeof(uuid_t));
```

Comment:
Juste a question: even if the UUID is not handled by the receiver, you send the message.
It is fine in our case as we need performances, however should it be possible to wait for an ACK to send the message ?

Reply:
It could be be done, but I am not in favor of changing the protocol in this PR.
Moreover, in case of data corruption, waiting for a ACK could lead to a dead-lock.
We could consider this change later if current implementation causes some problems in the future.